command/fmt_test.go

```diff
@@ -50,3 +53,109 @@ func TestFmt_unfomattedTemlateDirectory(t *testing.T) {
 		fatalCommand(t, c.Meta)
 	}
 }
+
+type RecursiveTestCase struct {
+	TestCaseName             string
+	Recursion                bool
+	TopLevelFilePreFormat    []byte
+	LowerLevelFilePreFormat  []byte
+	TopLevelFilePostFormat   []byte
+	LowerLevelFilePostFormat []byte
+}
+
+func TestFmt_Recursive(t *testing.T) {
+	unformattedData := []byte(`ami_filter_name ="amzn2-ami-hvm-*-x86_64-gp2"
+ami_filter_owners =[ "137112412989" ]
+
+`)
+
+	formattedData := []byte(`ami_filter_name   = "amzn2-ami-hvm-*-x86_64-gp2"
+ami_filter_owners = ["137112412989"]
+
+`)
+
+	recursiveTestCases := []RecursiveTestCase{
+		{
+			TestCaseName:             "With Recursive flag on",
+			Recursion:                true,
+			TopLevelFilePreFormat:    unformattedData,
+			LowerLevelFilePreFormat:  unformattedData,
+			TopLevelFilePostFormat:   formattedData,
+			LowerLevelFilePostFormat: formattedData,
+		},
+		{
+			TestCaseName:             "With Recursive flag off",
+			Recursion:                false,
+			TopLevelFilePreFormat:    unformattedData,
+			LowerLevelFilePreFormat:  unformattedData,
+			TopLevelFilePostFormat:   formattedData,
+			LowerLevelFilePostFormat: unformattedData,
+		},
```

Comment:
Nice one, but here I think that the map approach will be better, because I can choose and see directly from the test whatever levels of depths from the test is used, here's an example:

```go

	tests := []struct {
		name            string
		formatDir       string
		writeContent    map[string]string
		expectedContent map[string]string
	}{}

	for _, tt := range tests {
		topDir, err := ioutil.TempDir("", "")
		if err != nil {
			t.Fatalf("TempDir: %v", err)
		}
		for testDir, content := range tt.writeContent {
			ioutil.WriteFile(filepath.Join(topDir, testDir), []byte(content), 0666)
		}

		// run format on tt.formatDir

		for expectedPath, expectedContent := range tt.expectedContent {
			b, err := ioutil.ReadFile(filepath.Join(topDir, expectedPath))
			if err != nil {
				t.Fatalf("ReadFile: %v", err)
			}
			got := string(b)
			if diff := cmp.Diff(got, expectedContent); diff != "" {
				t.Errorf("format dir, unexpected %s: %s", expectedPath,)
			}
		}

	}
```

Reply:
While trying to implement I'm running into an issue with the key values in `alreadyPresentContent` and `expectedContent` because using the `os.TempDir` and `os.TempFile` functions, you cannot create multiple directories at once such as `foo/bar` at once using either of these functions. Because of this, it makes it very difficult to implement this as is with *n* level sub directories using the format you provided. Is there another approach I should try or am I missing something? Please let me know how to proceed.

Reply:
I would only create the top directory, and then use [`os.MkdirAll`]( https://golang.org/pkg/os/#MkdirAll) with the `foo/bar` paths, MkdirAll creates a directory named path, along with any necessary parents.

Reply:
The issue I'm facing with `os.MkdirAll` inside of a temp Directory is that it then doesn't allow for the tempDirectory to be cleaned up after the test function has finished (tested this without writing files inside of the os.MkdirAll directories as as with writing files). I could write some logic to create the temp directories if not already created, but I want to confirm with you this is the approach I should go with before implementing as this would greatly increase the complexity of this test case.

Reply:
Attempt at implementing is [here on this branch](https://github.com/teddylear/packer/blob/feature/recursivefmt-2-more-test-refactor/command/fmt_test.go#L67-L182), but this code leaves behind the test directory due to `os.MkdirAll`

Reply:
Gotcha, let me contribute to your PR by just changing the tests then ! Will be easier.

Reply:
Hey @teddylear, os.RemoveAll should get you sorted too. Also please merge master back into your branch; I think there are some conflicts.

If you want I can contribute to your branch also ?

Reply:
@Azr I believed I resolved the issue and have the tests working, but they are failing on windows for some reason on windows. I'm not sure why my CircleCI tests are cancelled either.

Reply:
Please feel free to contribute to my branch.

Reply:
@azr I was able to resolve issue with test not working on windows. Please let me know if there's anything else I should change.

Reply:
Also please note I believe I have addressed the requested change.